crypto/evp/keymgmt_lib.c

```diff
@@ -0,0 +1,289 @@
+#include "internal/provider.h"
+#include "evp_locl.h"
+
+static OSSL_PARAM *paramdefs_to_params(const OSSL_PARAM *paramdefs)
+{
+    size_t cnt;
+    const OSSL_PARAM *p;
+    OSSL_PARAM *params, *q;
+
+    for (cnt = 0, p = paramdefs; p->key != NULL; p++, cnt++)
+        continue;
+
+    params = OPENSSL_zalloc(cnt * sizeof(*params));
+
+    for (p = paramdefs, q = params; ; p++, q++) {
+        *q = *p;
+        if (p->key == NULL)
+            break;
+
+        q->data = NULL;          /* In case the provider used it */
+        q->return_size = 0;
+    }
+
+    return params;
+}
+
+typedef union align_block_un {
+    OSSL_UNION_ALIGN;
+} ALIGN_BLOCK;
+
+#define ALIGN_SIZE  sizeof(ALIGN_BLOCK)
+
+static void *allocate_params_space(OSSL_PARAM *params)
+{
+    unsigned char *data = NULL;
+    size_t space;
+    OSSL_PARAM *p;
+
+    for (space = 0, p = params; p->key != NULL; p++)
+        space += ((p->return_size + ALIGN_SIZE - 1) / ALIGN_SIZE) * ALIGN_SIZE;
+
+    data = OPENSSL_zalloc(space);
+
+    for (space = 0, p = params; p->key != NULL; p++) {
+        p->data = data + space;
+        space += ((p->return_size + ALIGN_SIZE - 1) / ALIGN_SIZE) * ALIGN_SIZE;
+    }
+
+    return data;
+}
+
+void *evp_keymgmt_export_to_provider(EVP_PKEY *pk, EVP_KEYMGMT *keymgmt)
+{
+    size_t i;
+
+    /*
+     * If there is an underlying legacy key and it has changed, invalidate
+     * the cache of provider keys.
+     */
+    if (pk->pkey.ptr != NULL) {
+        /*
+         * If there is no dirty checker, this key can't be used
+         * with providers
+         */
+        if (pk->ameth->is_dirty == NULL)
+            return NULL;
+
+        if (pk->ameth->is_dirty(pk))
+            evp_keymgmt_clear_pkey_cache(pk);
+    }
+
+    /*
+     * See if we have exported to this provider already.
+     * If we have, return immediately.
+     */
+    for (i = 0;
+         i < OSSL_NELEM(pk->pkeys) && pk->pkeys[i].keymgmt != NULL;
+         i++) {
+        if (keymgmt == pk->pkeys[i].keymgmt)
+            return pk->pkeys[i].provkey;
+    }
+
+    /*
+     * If there are no cached provider keys, try to export a legacy key
+     * to the provider.
+     */
+    if (i == 0) {
+        /* If there is no legacy key, give up */
+        if (pk->pkey.ptr == NULL)
+            return NULL;
+
+        /* Also, if there is no export function, give up */
+        if (pk->ameth->export_to == NULL)
+            return NULL;
+
+        /* Otherwise, simply use it and return the result */
+        EVP_KEYMGMT_up_ref(keymgmt);
+        pk->pkeys[0].keymgmt = keymgmt;
+        pk->pkeys[0].provkey = pk->ameth->export_to(pk, keymgmt);
+        return pk->pkeys[0].provkey;
+    }
+
+    /*
+     * There are cached provider keys, so import from the first one
+     * and export the imported data to the new provider.
+     * If it's impossible to import from the cached provider, give up
+     * (transfering keys between providers can't be supported in that
+     * case)
```

Comment:
I'm getting kind of confused about which direction we are talking about with export/import and from which perspective we are talking about it. i.e. does "import" mean import into a provider, or do we "import" a key from a provider into libcrypto. Above you talk about "export...to the new provider", but below you call the "importkey_priv" function.

Reply:
Hahaha, well, who's importing and who's exporting, that's the question, right?
So if the U.K. exports their Christmas Pudding, does Sweden import it or export it?
So on the same line of thinking, if libcrypto is exporting a key to a provider, is the provider importing or exporting it?